alphadia/workflow/optimization.py

```diff
@@ -579,3 +575,152 @@ def __init__(
+
+        Parameters
+        ----------
+        workflow: peptidecentric.PeptideCentricWorkflow
+            The workflow object, which includes the spectral library and the calibration and optimization managers.
+
+        """
+        self.workflow = workflow
+
+        self.reached_target_before = False
+        self.reached_target = False
+
+        self.elution_group_order = self.workflow.spectral_library.precursor_df[
+            "elution_group_idx"
+        ].unique()
+        np.random.shuffle(self.elution_group_order)
+
+        self.target_count = self.workflow.config["calibration"][
+            "optimization_lock_target"
+        ]
+
+        self.batch_idx = 0
+        self.set_batch_plan()
+
+        eg_idxes = self.elution_group_order[self.start_index : self.stop_index]
+        self.batch_precursor_df = self.workflow.spectral_library._precursor_df[
+            self.workflow.spectral_library._precursor_df["elution_group_idx"].isin(
+                eg_idxes
+            )
+        ]
+
+        self.total_precursors = len(self.batch_precursor_df)
+
+        self.features = []
+        self.fragments = []
+
+    def _get_exponential_batches(self, step):
+        """Get the number of batches for a given step
+        This plan has the shape:
+        1, 2, 4, 8, 16, 32, 64, ...
+        """
+        return int(2**step)
+
+    def set_batch_plan(self):
+        """Gets an exponential batch plan based on the batch_size value in the config."""
+        n_eg = self.workflow.spectral_library._precursor_df[
+            "elution_group_idx"
+        ].nunique()
+
+        plan = []
+
+        batch_size = self.workflow.config["calibration"]["batch_size"]
+        step = 0
+        start_index = 0
+
+        while start_index < n_eg:
+            n_batches = self._get_exponential_batches(step)
+            stop_index = min(start_index + n_batches * batch_size, n_eg)
+            plan.append((start_index, stop_index))
+            step += 1
+            start_index = stop_index
+
+        self.batch_plan = plan
+
+    def extract(self):
+        """Extract features and fragments from the current batch of the optimization lock."""
+
+        feature_df, fragment_df = self.workflow.extract_batch(self.batch_precursor_df)
+        self.features += [feature_df]
+        self.fragments += [fragment_df]
+
+    def fdr(self):
+        """Calculates the number of precursors at 1% FDR for the current optimization lock and determines if it is sufficient to perform calibration and optimization."""
+        self.precursor_df = self.workflow.fdr_correction(
+            self.features_df,
+            self.fragments_df,
+            self.workflow.optimization_manager.classifier_version,
+        )
+
+        self.count = len(self.precursor_df[self.precursor_df["qval"] < 0.01])
+
+        if self.count < self.target_count:
+            self.reached_target = False
+        else:
+            self.reached_target = True
+
+    def adjust_upwards(self):
```

Comment:
Do `adjust_upwards` and `reset_downwards` do the same just in opposite directions? 
If so, could we rename them to make this more clear?